helpers/utils.js

```diff
@@ -171,11 +171,19 @@ exports.performAction = async function performAction(request, action) {
             await page.setExtraHTTPHeaders(extraHeaders);
         }
 
+        const contextId = page.browserContext().id;
+        const pageId = page.target()._targetId;
+
         try {
-            return await action(page, request);
+            const response = await action(page, request);
+            return {
+                ...response,
+                contextId,
+                pageId,
+            };
         } catch (err) {
-            err.contextId = page.browserContext().id;
-            err.pageId = page.target()._targetId;
+            err.contextId = contextId;
+            err.pageId = pageId;
```

Comment:
Here I fix `action` requests from user.
Without it user does not get correct `scrapy-puppeteer-service-context-id` header (see routes/action.js).

Reply:
Ok
Do you think we should let user override page id (i.e. reversing to `{contextId, pageId, ...response}`)? It is technically possible to open new page in custom actions. We don't have real usages of it though and we may discourage it.

Reply:
Probably we should ban user from direct return `contextId` and `pageId` from the function as we override them in `performAction`.
Another solution I thought of is to return `{contextId, pageId, response}`. In this case we do not ban user from anything but we need to rewrite some parts in Python.